File: web/src/lib/colorDetect.js
// Two-stage dominant-colour detection for a vehicle crop.
//
// The old single-stage approach assigned a hue to every weakly-saturated pixel,
// so a white/silver car (which reflects blue sky and has a faint cast) could be
// voted "Blue". This version first decides whether the region is ACHROMATIC
// (white / silver / grey / black) from the *median saturation* of the body, and
// only classifies a hue when the region is genuinely colourful. That removes the
// white→blue error while keeping real colours accurate.

function rgbToHsv(r, g, b) {
  r /= 255; g /= 255; b /= 255;
  const max = Math.max(r, g, b), min = Math.min(r, g, b);
  const d = max - min;
  let h = 0;
  if (d !== 0) {
    if (max === r) h = ((g - b) / d) % 6;
    else if (max === g) h = (b - r) / d + 2;
    else h = (r - g) / d + 4;
    h *= 60;
    if (h < 0) h += 360;
  }
  return [h, max === 0 ? 0 : d / max, max];
}

function median(arr) {
  if (arr.length === 0) return 0;
  const a = arr.slice().sort((x, y) => x - y);
  const m = a.length >> 1;
  return a.length % 2 ? a[m] : (a[m - 1] + a[m]) / 2;
}

function hueName(h) {
  if (h < 15 || h >= 345) return "Red";
  if (h < 40) return "Orange";
  if (h < 70) return "Yellow";
  if (h < 160) return "Green";
  if (h < 200) return "Cyan";
  if (h < 255) return "Blue";
  if (h < 290) return "Purple";
  return "Magenta";
}

function shade(base, v) {
  if (v < 0.5) {
    if (base === "Red") return "Maroon";
    if (base === "Blue") return "Navy";
    if (base === "Green") return "Dark Green";
    if (base === "Yellow") return "Olive";
    if (base === "Orange") return "Brown";
    if (base === "Purple") return "Indigo";
  }
  return base;
}

function hex(r, g, b) {
  return (
    "#" +
    [r, g, b].map((x) => Math.round(x).toString(16).padStart(2, "0")).join("")
  );
}

// Tunables.
const GRAY_S = 0.25;       // below this median saturation -> achromatic
const SAT_GATE = 0.32;     // a pixel must beat this saturation to vote a hue
const MIN_SAT_FRAC = 0.25; // need this fraction of saturated pixels to be chromatic
const MIN_WIN_SAT = 0.40;  // winning hue must be this saturated, else it's a tint

function achromatic(medV) {
  if (medV < 0.22) return "Black";
  if (medV > 0.70) return "White";
  if (medV > 0.42) return "Silver/Gray";
  return "Dark Gray";
}

export function detectColor(ctx, box) {
  const cw = ctx.canvas.width, ch = ctx.canvas.height;

  // Body region: lower-centre, avoiding glass and edges.
  let rx = Math.round(box.x + box.w * 0.22);
  let ry = Math.round(box.y + box.h * 0.45);
  let rw = Math.round(box.w * 0.56);
  let rh = Math.round(box.h * 0.45);
  rx = Math.max(0, Math.min(rx, cw - 2));
  ry = Math.max(0, Math.min(ry, ch - 2));
  rw = Math.max(1, Math.min(rw, cw - rx));
  rh = Math.max(1, Math.min(rh, ch - ry));

  let data;
  try {
    data = ctx.getImageData(rx, ry, rw, rh).data;
  } catch (e) {
    return { name: "Unknown", hex: "#888888", confidence: 0 };
  }

  const H = [], S = [], V = [], R = [], G = [], B = [];
  const step = Math.max(1, Math.floor((rw * rh) / 2000)) * 4;
  for (let i = 0; i < data.length; i += step) {
    const r = data[i], g = data[i + 1], b = data[i + 2];
    const [h, s, v] = rgbToHsv(r, g, b);
    H.push(h); S.push(s); V.push(v); R.push(r); G.push(g); B.push(b);
  }
  const n = S.length;
  if (n === 0) return { name: "Unknown", hex: "#888888", confidence: 0 };

  const medS = median(S);
  const medV = median(V);

  let satCount = 0;
  for (let i = 0; i < n; i++) if (S[i] > SAT_GATE && V[i] > 0.2) satCount++;
  const satFrac = satCount / n;

  const avgAll = () => {
    let rs = 0, gs = 0, bs = 0;
    for (let i = 0; i < n; i++) { rs += R[i]; gs += G[i]; bs += B[i]; }
    return hex(rs / n, gs / n, bs / n);
  };

  // ---- Stage 1: achromatic (white / silver / grey / black) ----
  if (medS < GRAY_S || satFrac < MIN_SAT_FRAC) {
    return { name: achromatic(medV), hex: avgAll(), confidence: Math.round((1 - medS) * 100) };
  }

  // ---- Stage 2: chromatic — saturation-WEIGHTED vote over saturated pixels ----
  const weight = {};   // name -> summed saturation (vote weight)
  const sums = {};     // name -> [r,g,b,count,satSum]
  for (let i = 0; i < n; i++) {
    if (S[i] <= SAT_GATE || V[i] <= 0.2) continue;
    const name = hueName(H[i]);
    weight[name] = (weight[name] || 0) + S[i];
    const acc = sums[name] || (sums[name] = [0, 0, 0, 0, 0]);
    acc[0] += R[i]; acc[1] += G[i]; acc[2] += B[i]; acc[3]++; acc[4] += S[i];
  }
  let best = null, bestW = -1;
  for (const k in weight) if (weight[k] > bestW) { bestW = weight[k]; best = k; }
  if (!best) {
    return { name: achromatic(medV), hex: avgAll(), confidence: Math.round((1 - medS) * 100) };
  }
  const acc = sums[best];
  const winMeanSat = acc[4] / acc[3];

  // Guard: if the "winning" colour is actually a weak tint (e.g. sky reflection
  // on a white car), treat the vehicle as achromatic instead.
  if (winMeanSat < MIN_WIN_SAT) {
    return { name: achromatic(medV), hex: avgAll(), confidence: Math.round((1 - winMeanSat) * 100) };
  }

  return {
    name: shade(best, medV),
    hex: hex(acc[0] / acc[3], acc[1] / acc[3], acc[2] / acc[3]),
    confidence: Math.round(Math.min(1, winMeanSat) * 100),
  };
}

File: python/vd/color_detect.py
"""Two-stage dominant-colour detection for a vehicle crop (OpenCV/NumPy).

Stage 1 decides whether the vehicle body is ACHROMATIC (white / silver / grey /
black) using the *median saturation* of the region. Only if the region is
genuinely colourful does Stage 2 classify a hue (using saturated pixels only).
This removes the common white→blue error caused by sky reflections on light cars.
"""

import cv2
import numpy as np

GRAY_S = 0.25        # below this median saturation -> achromatic
SAT_GATE = 0.32      # a pixel must beat this saturation to vote a hue
MIN_SAT_FRAC = 0.25  # need this fraction of saturated pixels to be chromatic
MIN_WIN_SAT = 0.40   # winning hue must be this saturated, else it's a tint


def _achromatic(med_v):
    if med_v < 0.22:
        return "Black"
    if med_v > 0.70:
        return "White"
    if med_v > 0.42:
        return "Silver/Gray"
    return "Dark Gray"


def _hue_base(h):
    """Vectorised hue (0-360) -> base-colour code 0..7."""
    conds = [
        (h < 15) | (h >= 345),   # 0 Red
        (h >= 15) & (h < 40),    # 1 Orange
        (h >= 40) & (h < 70),    # 2 Yellow
        (h >= 70) & (h < 160),   # 3 Green
        (h >= 160) & (h < 200),  # 4 Cyan
        (h >= 200) & (h < 255),  # 5 Blue
        (h >= 255) & (h < 290),  # 6 Purple
    ]
    return np.select(conds, [0, 1, 2, 3, 4, 5, 6], default=7)  # 7 Magenta


BASE_NAMES = ["Red", "Orange", "Yellow", "Green", "Cyan", "Blue", "Purple", "Magenta"]


def _shade(base, v):
    if v < 0.5:
        return {
            "Red": "Maroon", "Blue": "Navy", "Green": "Dark Green",
            "Yellow": "Olive", "Orange": "Brown", "Purple": "Indigo",
        }.get(base, base)
    return base


def detect_color(region_bgr):
    """Return (name, hex) for a BGR vehicle crop."""
    if region_bgr is None or region_bgr.size == 0:
        return "Unknown", "#888888"
    h, w = region_bgr.shape[:2]
    if h < 6 or w < 6:
        return "Unknown", "#888888"

    y0, y1 = int(h * 0.45), int(h * 0.90)
    x0, x1 = int(w * 0.22), int(w * 0.78)
    crop = region_bgr[y0:y1, x0:x1]
    if crop.size == 0:
        crop = region_bgr

    hsv = cv2.cvtColor(crop, cv2.COLOR_BGR2HSV)
    H = hsv[..., 0].astype(np.float32).ravel() * 2.0     # 0-360
    S = hsv[..., 1].astype(np.float32).ravel() / 255.0
    V = hsv[..., 2].astype(np.float32).ravel() / 255.0
    flat = crop.reshape(-1, 3)                            # BGR

    med_s = float(np.median(S))
    med_v = float(np.median(V))
    sat_mask = (S > SAT_GATE) & (V > 0.2)
    sat_frac = float(sat_mask.mean())

    def hexof(arr):
        b, g, r = int(arr[0]), int(arr[1]), int(arr[2])
        return "#%02x%02x%02x" % (r, g, b)

    # ---- Stage 1: achromatic ----
    if med_s < GRAY_S or sat_frac < MIN_SAT_FRAC:
        return _achromatic(med_v), hexof(flat.mean(axis=0))

    # ---- Stage 2: chromatic — saturation-WEIGHTED vote over saturated pixels ----
    codes = _hue_base(H[sat_mask]).astype(np.int32)
    sat_vals = S[sat_mask]
    weights = np.bincount(codes, weights=sat_vals, minlength=len(BASE_NAMES))
    best = int(np.argmax(weights))

    sat_flat = flat[sat_mask]
    chosen = sat_flat[codes == best]
    win_mean_sat = float(sat_vals[codes == best].mean()) if len(chosen) else 0.0

    # Guard: weak tint (e.g. sky reflection on a white car) -> achromatic.
    if win_mean_sat < MIN_WIN_SAT:
        return _achromatic(med_v), hexof(flat.mean(axis=0))

    name = _shade(BASE_NAMES[best], med_v)
    avg = chosen.mean(axis=0) if len(chosen) else sat_flat.mean(axis=0)
    return name, hexof(avg)
